dvc/remote/s3.py

```diff
@@ -180,27 +184,57 @@ def remove(self, path_info):
         logger.debug("Removing {}".format(path_info))
         self.s3.delete_object(Bucket=path_info.bucket, Key=path_info.path)
 
-    def _list_paths(self, bucket, prefix):
+    def _list_objects(self, path_info, max_items=None):
         """ Read config for list object api, paginate through list objects."""
-        kwargs = {"Bucket": bucket, "Prefix": prefix}
-        if self.list_objects:
-            list_objects_api = "list_objects"
-        else:
-            list_objects_api = "list_objects_v2"
-        paginator = self.s3.get_paginator(list_objects_api)
+        kwargs = {
+            "Bucket": path_info.bucket,
+            "Prefix": path_info.path,
+            "PaginationConfig": {"MaxItems": max_items},
+        }
+        paginator = self.s3.get_paginator(self.list_objects_api)
         for page in paginator.paginate(**kwargs):
             contents = page.get("Contents", None)
             if not contents:
                 continue
             for item in contents:
-                yield item["Key"]
+                yield item
+
+    def _list_paths(self, path_info, max_items=None):
+        return (
+            item["Key"] for item in self._list_objects(path_info, max_items)
+        )
 
     def list_cache_paths(self):
-        return self._list_paths(self.path_info.bucket, self.path_info.path)
+        return self.walk_files(self.path_info)
 
     def exists(self, path_info):
-        paths = self._list_paths(path_info.bucket, path_info.path)
-        return any(path_info.path == path for path in paths)
+        dir_path = path_info / ""
+        fname = next(self._list_paths(path_info, max_items=1), "")
+        return path_info.path == fname or fname.startswith(dir_path.path)
+
+    def isdir(self, path_info):
+        # S3 doesn't have a concept for directories.
+        #
+        # Using `head_object` with a path pointing to a directory
+        # will throw a 404 error.
+        #
+        # A reliable way to know if a given path is a directory is by
+        # checking if there are more files sharing the same prefix
+        # with a `list_objects` call.
+        #
+        # We need to make sure that the path ends with a forward slash,
+        # since we can end with false-positives like the following example:
+        #
+        #       bucket
+        #       └── data
```

Comment:
On python2:
```
+python -mtests
ImportError while loading conftest '/home/travis/build/iterative/dvc/tests/conftest.py'.
tests/conftest.py:8: in <module>
    from dvc.remote.config import RemoteConfig
dvc/remote/__init__.py:7: in <module>
    from dvc.remote.s3 import RemoteS3
E     File "/home/travis/build/iterative/dvc/dvc/remote/s3.py", line 229
E   SyntaxError: Non-ASCII character '\xe2' in file /home/travis/build/iterative/dvc/dvc/remote/s3.py on line 229, but no encoding declared; see http://python.org/dev/peps/pep-0263/ for details
Traceback (most recent call last):
  File "/opt/python/2.7.15/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/opt/python/2.7.15/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/travis/build/iterative/dvc/tests/__main__.py", line 20, in <module>
    check_call(cmd, shell=True)
  File "/opt/python/2.7.15/lib/python2.7/subprocess.py", line 190, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'py.test -v -n=4 --timeout=600 --timeout_method=thread --log-level=debug --cov=dvc  --durations=0' returned non-zero exit status 4
The command "bash ./scripts/ci/script.sh" exited with 1.
cache.2
store build cache
```

Reply:
ohh, right, leaving comments with unicode is broken :see_no_evil: